pkg/util/provider/machinecontroller/machine_util.go

```diff
@@ -708,31 +718,78 @@ func (c *controller) reconcileMachineHealth(ctx context.Context, machine *v1alph
 					machine.Name,
 					timeOutDuration,
 				)
+				// Log the error message for machine failure
+				klog.Error(description)
+
+				clone.Status.LastOperation = v1alpha1.LastOperation{
+					Description:    description,
+					State:          v1alpha1.MachineStateFailed,
+					Type:           machine.Status.LastOperation.Type,
+					LastUpdateTime: metav1.Now(),
+				}
+				clone.Status.CurrentStatus = v1alpha1.CurrentStatus{
+					Phase: v1alpha1.MachineFailed,
+					//TimeoutActive:  false,
+					LastUpdateTime: metav1.Now(),
+				}
+				objectRequiresUpdate = true
 			} else {
-				// Timeour occurred due to machine being unhealthy for too long
+
+				// Timeout occurred due to machine being unhealthy for too long
 				description = fmt.Sprintf(
 					"Machine %s is not healthy since %s minutes. Changing status to failed. Node Conditions: %+v",
 					machine.Name,
 					timeOutDuration,
 					machine.Status.Conditions,
 				)
-			}
 
-			// Log the error message for machine failure
-			klog.Error(description)
+				machineSetName, machineDeployName := getMachineOwnerNames(machine)
+				//getting lock for machineDeployment this machine belongs to, creating lock if absent
+				tempCh := make(chan struct{}, 1)
+				val, loaded := c.safetyOptions.HealthChanMap.LoadOrStore(machineDeployName, tempCh)
+				if loaded {
+					close(tempCh)
```

Comment:
any reason that you close the channel if loaded is true?

Reply:
`HealthChanMap` is a map with key as machineDeployment name and value as the lock/channel for that machineDeployment. In these line we try to create entry for the machineDeployment if not there in the map. If the entry is there then I delete the tempCh (which I created to make entry in the map for that machineDeployment).